src/chunk_index.c

```diff
@@ -0,0 +1,533 @@
+#include <postgres.h>
+#include <catalog/index.h>
+#include <catalog/indexing.h>
+#include <catalog/pg_index.h>
+#include <catalog/pg_constraint.h>
+#include <catalog/objectaddress.h>
+#include <catalog/namespace.h>
+#include <access/htup_details.h>
+#include <utils/syscache.h>
+#include <utils/lsyscache.h>
+#include <utils/rel.h>
+#include <utils/fmgroids.h>
+#include <utils/builtins.h>
+#include <nodes/parsenodes.h>
+#include <commands/defrem.h>
+#include <commands/tablecmds.h>
+
+#include "chunk_index.h"
+#include "hypertable.h"
+#include "hypertable_cache.h"
+#include "catalog.h"
+#include "scanner.h"
+#include "chunk.h"
+
+static List *
+create_index_colnames(Relation indexrel)
+{
+	List	   *colnames = NIL;
+	int			i;
+
+	for (i = 0; i < indexrel->rd_att->natts; i++)
+		colnames = lappend(colnames, pstrdup(NameStr(indexrel->rd_att->attrs[i]->attname)));
+
+	return colnames;
+}
+
+/*
+ * Get a list of a relation's non-constraint indexes.
+ *
+ * Constraint indexes are created as a result of creating the associated
+ * constraint, so we want to exclude them when we create indexes on a chunk.
+ */
+static List *
+relation_get_non_constraint_indexes(Relation rel)
+{
+	List	   *indexlist = RelationGetIndexList(rel);
+	Relation	conrel;
+	SysScanDesc conscan;
+	ScanKeyData skey[1];
+	HeapTuple	htup;
+
+	ScanKeyInit(&skey[0],
+				Anum_pg_constraint_conrelid,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(RelationGetRelid(rel)));
+
+	conrel = heap_open(ConstraintRelationId, AccessShareLock);
+	conscan = systable_beginscan(conrel, ConstraintRelidIndexId, true,
+								 NULL, 1, skey);
+
+	while (HeapTupleIsValid(htup = systable_getnext(conscan)))
+	{
+		Form_pg_constraint conform = (Form_pg_constraint) GETSTRUCT(htup);
+
+		indexlist = list_delete_oid(indexlist, ObjectIdGetDatum(conform->conindid));
+	}
+
+	systable_endscan(conscan);
+	heap_close(conrel, AccessShareLock);
+
+	return indexlist;
+}
+
+/*
+ * Pick a name for a chunk index.
+ *
+ * The goal is that a chunk index name should be analogous to the corresponding
+ * hypertable index name. For instance, if the hypertable's index name is
+ * "<table_name>_<columns>_<suffix>" we should generate the name
+ * "<chunk_table_name>_<columns>_<suffix>".
```

Comment:
If you are referring to your comment about "Unique1" vs <chunk_name>_<column_names> in a test output below, this function is not responsible for that output.

The reason you get that mismatch is because when you are creating a new index on an existing table, this function is actually not involved, and instead the IndexStmt is simply recursed to chunks, letting PostgreSQL take care of naming indexes and resolving conflicts. 

So, as I see it, there are two approaches to naming indexes:

1) we let PostgreSQL pick names to the extent possible, reusing its code.
2) we always generate our own names using our own code.

I opted for (1). Still, the problem is that the index naming functions are non-public, so the only way to get PostgreSQL to generate a name is to set index name to NULL in the IndexStmt for each chunk. Although this admittedly gives some not ideal corner-case mappings, but it does use PostgreSQL's own naming functions and conflict resolution to the extent possible. 

Alternative (2) probably means using this function or a variant, and maybe that is the approach we should take rather than letting PostgreSQL pick names? I am certainly open for it. Probably <chunk_table>_<original_index_name> is the simplest approach, with some conflict resolution in place. 

For the record though, the previous implementation didn't do any conflict resolution and simply threw an error trace if an index name conflicted on a chunk. Not a likely error, but still possible technically.  So, no matter the naming scheme we end up with this is an improvement from a correctness standpoint.